pkg/controller/eventinstance/eventinstance.go

```diff
@@ -106,3 +104,118 @@ func (h handler) gcExpired(
 
 	return nil
 }
+
+func GCOverCapacity(c kclient.Client, capacity int) router.HandlerFunc {
+	sentinel := &capacitySentinel{
+		client:   c,
+		capacity: capacity,
+	}
+	return func(req router.Request, _ router.Response) error {
+		e := req.Object.(*internalv1.EventInstance)
+		var err error
+		if e.DeletionTimestamp != nil && !e.DeletionTimestamp.IsZero() {
+			err = sentinel.rm(req.Ctx, *e)
+		} else {
+			err = sentinel.add(req.Ctx, *e)
+		}
+
+		return err
+	}
+}
+
+type capacitySentinel struct {
+	client      kclient.Client
+	capacity    int
+	initialized bool
+	once        sync.Once
+	sem         *semaphore.Weighted
+	tracking    map[types.UID]struct{}
+}
+
+func (s *capacitySentinel) add(ctx context.Context, e internalv1.EventInstance) error {
+	s.once.Do(func() {
+		s.sem = semaphore.NewWeighted(1)
```

Comment:
I'm going to switch to a regular `sync.Mutex`; there are bigger problems if the request context expires while we're waiting for a lock, and we'll see them as at a higher level.